Update default settings for non unique index sampling, sampler set initial size evaluation.

MishaDemianenko opened the pull request:
Decrease default non unique sampler step size limit to 8m instead of 64m.
Update default set size evaluation to be less greedy, make it sample step size dependent.

Performance wise new default size/initial set size combination gives about 30% increase while populating sampler.
And it does not make sense to create new map for values each time.

Closes #7106.

MishaDemianenko committed:
Update default settings for non unique index sampling, sampler set initial size evaluation.

Decrease default non unique sampler step size limit to 8m instead of 64m.
Update default set size evaluation to be less greedy, make it sample step size dependent.

Performance wise new default size/initial set size combination gives about 30% increase while populating sampler.
And it does not make sense to create new map for values each time.

MishaDemianenko commented:
@chrisvest i have lowered a bit default settings and did some measurements in regards of new map creating instead. Would propose we keep old property name in 3.0.x and rename and modify it a bit when we will merge it in 3.1 (or in separate commit).

@systay if this PR will find it's way in it will decrease default sampler step size for non unique indexes, just FYI

chrisvest commented:
@MishaDemianenko Strictly speaking, we need to add a migration for the changed configuration, and I think we might as well do that in 3.0 since it can be done without breaking any sort of compatibility.

chrisvest commented:
Apart from that, I like the changes.

MishaDemianenko committed:
Introduce new property dbms.index_sampling.sample_size_limit instead of old dbms.index_sampling.buffer_size.

Introduce new better named and integer property dbms.index_sampling.buffer_size.
Provide property migration as well.

MishaDemianenko commented:
@chrisvest new property created and migration provided 
@benbc we manipulate with settings here and introduce new `index_sampling.sample_size_limit` instead of `dbms.index_sampling.buffer_size` and provide migration.

benbc commented:
@MishaDemianenko Looks great. Thanks for showing me this.

chrisvest commented on community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java:
```diff
@@ -303,11 +303,18 @@
-    public static final Setting<Long> index_sampling_buffer_size =
-            setting("dbms.index_sampling.buffer_size", BYTES, "8m",
+    @Description("Size of buffer used by index sampling. " +
+                 "This configuration setting is no longet applicable as from Neo4j 3.0.3." +
```
`longer` instead of `longet`

MishaDemianenko replied:
@chrisvest  fixed

MishaDemianenko committed:
Fix mistype londet -> longer.

chrisvest commented:
Looks good. Will merge after 3.0.2 is out.
